main.tf

```
@@ -1,65 +1,88 @@
 }
 
-module "vpc_ipam_pool_cidr_allocation" {
-  source           = "./modules/pool-cidr-allocation"
-  cidr             = var.pool_cidr_allocation.cidr
-  disallowed_cidrs = var.pool_cidr_allocation.disallowed_cidrs
-  ipam_pool_id     = var.pool_cidr_allocation.ipam_pool_id
-  netmask_length   = var.pool_cidr_allocation.netmask_length
+
+#plural, provisions a CIDR from an IPAM pool
+#netmask conflisks with cidr and for_each statement
+resource "aws_vpc_ipam_pool_cidr" "this" {
+  for_each = { for cidr in var.vpc_ipam.pool_cidrs : cidr.cidr => cidr }
+
+
+  cidr = each.value.cidr
+  cidr_authorization_context {
+    message   = each.value.cidr_authorization_context_message
+    signature = each.value.cidr_authorization_context_signature
+  }
+  ipam_pool_id   = each.value.ipam_pool_id
+  netmask_length = each.value.netmask_length
 }
 
-module "vpc_ipam_resource_discovery" {
-  source      = "./modules/resource-discovery"
-  region_name = var.resource_discovery.region_name
-  tags        = var.resource_discovery.tags
+#plural, carves out a CIDR allocation from pool, preventing usage
+resource "aws_vpc_ipam_pool_cidr_allocation" "this" {
+  for_each = { for allocation in var.vpc_ipam.pool_cidr_allocations : allocation.cidr => allocation }
+
+  cidr             = each.value.cidr
+  description      = each.value.description
+  disallowed_cidrs = each.value.disallowed_cidrs
+  ipam_pool_id     = each.value.ipam_pool_id
+  netmask_length   = each.value.netmask_length
 }
 
-module "vpc_ipam_resource_discovery_association" {
-  source                     = "./modules/resource-discovery-association"
-  ipam_id                    = var.resource_discovery_association.ipam_id
-  ipam_resource_discovery_id = var.resource_discovery_association.ipam_resource_discovery_id
-  tags                       = var.resource_discovery_association.tags
+#for checking next available CIDR,
+# unsure of implementation usage? could be useful for checking/assigning next CIDR (dynamically)?
+resource "aws_vpc_ipam_preview_next_cidr" "this" {
+  count = local.create_vpc_ipam_preview_next_cidr && var.vpc_ipam.preview_next_cidr != null ? 1 : 0
+
+  #disallowed cidrs could be a dynamic input
+  disallowed_cidrs = var.vpc_ipam.preview_next_cidr.disallowed_cidrs != null ? var.vpc_ipam.preview_next_cidr.disallowed_cidrs : []
+  ipam_pool_id     = try(var.vpc_ipam.preview_next_cidr.ipam_pool_id, [])
+  netmask_length   = try(var.vpc_ipam.preview_next_cidr.netmask_length, [])
 }
 
-module "vpc_ipam_preview_next_cidr" {
-  source           = "./modules/preview-next-cidr"
-  disallowed_cidrs = var.preview_next_cidr.disallowed_cidrs
-  ipam_pool_id     = var.preview_next_cidr.ipam_pool_id
-  netmask_length   = var.preview_next_cidr.netmask_length
+#plural, scopes segment to multiple disconnected networks
+resource "aws_vpc_ipam_scope" "this" {
+  for_each = { for scope in var.vpc_ipam.scopes : scope.name => scope }
+
+  ipam_id     = each.value.ipam_id
```

It would be nice to be able to create the ipam and top-level scopes in one module call... We should think about how we might allow the user to choose to reference the attributes `aws_vpc_ipam.this.private_default_scope_id` or `aws_vpc_ipam.example.public_default_scope_id`, instead of requring them to provide the ipam_id...

---

Actually, the scopes can always be _this_ ipam. And we can use a precondition to fail if the user sets `var.vpc_ipam.scopes` but does not set `var.vpc_ipam_ipam`.

```suggestion
  ipam_id     = aws_vpc_ipam.this[0].ipam_id
```